### PR TITLE
Secure connection to Postgres

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Summary

The idea is to secure Postgres so we can safely transfer secrets between it and Kuma CP.
I introduced new config settings for corresponding Postgres options.
Integration tests now runs with Postgres which is secured in Verify CA mode.

The most challenging part as usual is generating certificates. To have proper security you need to generate Root CA and then Server and Client cert signing with this CA. For now I generated RootCA and two certs manually.

Maybe we need more options to `kumactl generate tls-certificate`? But on the other hand, when there is a CA in the organisation, don't they have automation to generate certs anyways?

### CODE REVIEW

File: tools/postgres/ssl/pg_hba.conf [hunk 1]
```diff
@@ -1,2 +1,3 @@
 # TYPE  DATABASE        USER            ADDRESS                 METHOD
+local	all				all										trust
```
Comment (yskopets) [inline tools/postgres/ssl/pg_hba.conf]: Why do we need this line ?

Documentation says `This record matches connection attempts using Unix-domain sockets. Without a record of this type, Unix-domain socket connections are disallowed.`

Comment (jakubdyszkiewicz) [reply tools/postgres/ssl/pg_hba.conf]: It's for initial script that is applied after docker compose is up

File: pkg/config/plugins/resources/postgres/config.go [hunk 2]
```diff
@@ -22,6 +21,58 @@ type PostgresStoreConfig struct {
...
+	switch s.Mode {
+	case VerifyFull:
+		fallthrough
+	case VerifyCa:
+		if s.RootCertPath == "" {
+			return errors.New("RootCertPath cannot be empty")
+		}
+		fallthrough
+	case Require:
+		if s.CertPath == "" {
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/config/plugins/resources/postgres/config.go]: It seems so. If you try to run this without certs you will get
```
pq: connection requires a valid client certificate
```

[the remark this answers is not in the document]
Comment (yskopets) [reply pkg/config/plugins/resources/postgres/config.go]: I think, certificate is required due to `auth-method` set to `cert` in `pg_hba.conf`, i.e.
```
hostssl all             all             0.0.0.0/0               cert
```
But end users might use configurations with other `auth-method`s

File: pkg/config/app/kuma-cp/kuma-cp.defaults.yaml [hunk 3]
```diff
@@ -25,9 +25,9 @@ store:
     dbName: kuma # ENV: KUMA_STORE_POSTGRES_DB_NAME
     # Connection Timeout to the DB in seconds
     connectionTimeout: 5 # ENV: KUMA_STORE_POSTGRES_CONNECTION_TIMEOUT
-    # SSL settings
-    ssl:
-      # Mode of SSL connection. Available values (disable, require, verify-ca, verify-full)
+    # TLS settings
+    tls:
+      # Mode of TLS connection. Available values (disable, require, verify-ca, verify-full)
       mode: disable # ENV: KUMA_STORE_POSTGRES_SSL_MODE
```
Comment (yskopets) [inline pkg/config/app/kuma-cp/kuma-cp.defaults.yaml]: Let's rename `SSL` leftovers in comments

File: pkg/config/plugins/resources/postgres/config.go [hunk 4]
```diff
@@ -109,17 +109,17 @@ func DefaultPostgresStoreConfig() *PostgresStoreConfig {
...
-var _ config.Config = &SSLPostgresStoreConfig{}
+var _ config.Config = &TLSPostgresStoreConfig{}
 
-func DefaultSSLPostgresStoreConfig() SSLPostgresStoreConfig {
-	return SSLPostgresStoreConfig{
-		Mode:         Disable,
-		CertPath:     "",
-		KeyPath:      "",
-		RootCertPath: "",
+func DefaultSSLPostgresStoreConfig() TLSPostgresStoreConfig {
```
Comment (yskopets) [inline pkg/config/plugins/resources/postgres/config.go]: Let's rename `SSL` leftovers

File: pkg/config/app/kuma-cp/kuma-cp.defaults.yaml [hunk 5]
```diff
@@ -25,6 +25,16 @@ store:
...
+    # TLS settings
+    tls:
+      # Mode of TLS connection. Available values (disable, require, verify-ca, verify-full)
+      mode: disable # ENV: KUMA_STORE_POSTGRES_SSL_MODE
+      # Path to SSL Certificate of the client. Used in require, verify-ca and verify-full modes
+      certPath: # ENV: KUMA_STORE_POSTGRES_SSL_CERT_PATH
+      # Path to SSL Key of the client. Used in require, verify-ca and verify-full modes
+      keyPath: # ENV: KUMA_STORE_POSTGRES_SSL_KEY_PATH
+      # Path to the root certificate. Used in verify-ca and verify-full modes.
+      rootCertPath: # ENV: KUMA_STORE_POSTGRES_SSL_ROOT_CERT_PATH
```
Comment (yskopets) [inline pkg/config/app/kuma-cp/kuma-cp.defaults.yaml]: Still called `rootCertPath` here

File: tools/postgres/ssl/docker-compose.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    environment:
+      POSTGRES_USER: ${KUMA_STORE_POSTGRES_USER:-kuma}
+      POSTGRES_PASSWORD: ${KUMA_STORE_POSTGRES_PASSWORD:-kuma}
+      POSTGRES_DB: ${KUMA_STORE_POSTGRES_DB_NAME:-kuma}
+    ports:
+      - ${KUMA_STORE_POSTGRES_PORT:-15432}:5432
+    volumes:
+      - ../../../app/kumactl/cmd/install/testdata/postgres_schema.golden.sql:/docker-entrypoint-initdb.d/init.sql
+      - ./certs/rootCA.crt:/var/lib/postgresql/rootCA.crt:ro
+      - ./certs/postgres.server.crt:/var/lib/postgresql/postgres.server.crt:ro
```
Comment (yskopets) [inline tools/postgres/ssl/docker-compose.yaml]: Why we are not mounting server's TLS key ? (`/var/lib/postgresql/postgres.server.key`)

File: pkg/config/plugins/resources/postgres/config.go [hunk 7]
```diff
@@ -22,6 +21,57 @@ type PostgresStoreConfig struct {
...
+	case VerifyFull:
+		fallthrough
+	case VerifyCa:
+		if s.CAPath == "" {
+			return errors.New("CAPath cannot be empty")
+		}
+		if s.CertPath == "" {
+			return errors.New("CertPath cannot be empty")
+		}
+		if s.KeyPath == "" {
```
Comment (yskopets) [inline pkg/config/plugins/resources/postgres/config.go]: Why do we still require client TLS cert and key ?

File: tools/postgres/ssl/docker-compose.yaml (hunk 6 above)
Comment (jakubdyszkiewicz) [reply tools/postgres/ssl/docker-compose.yaml]: Unfortunately, I could not find a way to mount it with proper permissions

File: pkg/config/plugins/resources/postgres/config.go [hunk 8]
```diff
@@ -22,6 +21,51 @@ type PostgresStoreConfig struct {
...
+	case VerifyCa:
+		if s.CAPath == "" {
+			return errors.New("CAPath cannot be empty")
+		}
+	case VerifyNone:
+	case Disable:
+	default:
+		return errors.Errorf("invalid mode: %s", s.Mode)
+	}
+	return nil
```
Comment (yskopets) [inline pkg/config/plugins/resources/postgres/config.go]: Let's verify that if `Cert` is non-empty then so must be the `Key`. And vice versa

File: tools/postgres/ssl/docker-compose.yaml (hunk 6 above)
Comment (yskopets) [reply tools/postgres/ssl/docker-compose.yaml]: Let's add a note about it next to this line, and mention how it is solved instead.

**Pull request merged**